config/initializers/telemetry.rb

```diff
@@ -0,0 +1,87 @@
+    StatsD.backend = StatsD::Instrument::Backends::UDPBackend.new(target, SETTINGS[:telemetry_statsd_protocol].to_s)
+  else
+    StatsD.backend = StatsD::Instrument::Backends::NullBackend.new
+  end
+end
+StatsD.prefix = SETTINGS[:telemetry_prefix] || "foreman.#{Socket.gethostname}"
+
+module Foreman::Telemetry
+  def self.measure(key, value = nil, *metric_options, &block)
+    if value.is_a?(Hash) && metric_options.empty?
+      metric_options = [value]
+      value = value.fetch(:value, nil)
+    end
+
+    result = nil
+    value  = 1000 * StatsD::Instrument.duration { result = block.call } if block_given?
+    ActiveSupport::Notifications.instrument 'telemetry.measure', :key => key, :value => value, :metric_options => metric_options
+    result
+  end
+
+  def self.increment(key, value = 1, *metric_options)
+    if value.is_a?(Hash) && metric_options.empty?
+      metric_options = [value]
+      value = value.fetch(:value, 1)
+    end
+
+    ActiveSupport::Notifications.instrument 'telemetry.increment', :key => key, :value => value, :metric_options => metric_options
+  end
+
+  def self.gauge(key, value, *metric_options)
+    if value.is_a?(Hash) && metric_options.empty?
+      metric_options = [value]
+      value = value.fetch(:value, nil)
+    end
+
+    ActiveSupport::Notifications.instrument 'telemetry.gauge', :key => key, :value => value, :metric_options => metric_options
+  end
+end
+
+ActiveSupport::Notifications.subscribe /telemetry.measure/ do |name, start, finish, id, payload|
+  StatsD.measure payload[:key], payload[:value], payload[:metric_options] || {}
+end
+
+ActiveSupport::Notifications.subscribe /telemetry.increment/ do |name, start, finish, id, payload|
+  StatsD.increment payload[:key], payload[:value] || 1, payload[:metric_options] || {}
+end
+
+ActiveSupport::Notifications.subscribe /telemetry.gauge/ do |name, start, finish, id, payload|
+  StatsD.gauge payload[:key], payload[:value], payload[:metric_options] || {}
+end
+
+ActiveSupport::Notifications.subscribe /process_action.action_controller/ do |*args|
+  event = ActiveSupport::Notifications::Event.new(*args)
+  controller = event.payload[:controller]
+  action = event.payload[:action]
+  format = event.payload[:format] || "all"
+  format = "all" if format == "*/*"
+  status = event.payload[:status]
+  key = "#{controller}.#{action}.#{format}"
+  Foreman::Telemetry.measure "#{key}.total_duration", event.duration
+  Foreman::Telemetry.measure "#{key}.db_time", event.payload[:db_runtime]
+  Foreman::Telemetry.measure "#{key}.view_time", (event.payload[:view_runtime] || 0)
+  Foreman::Telemetry.increment "#{key}.status.#{status}"
+end
+
+ActiveSupport::Notifications.subscribe /instantiation.active_record/ do |*args|
+  event = ActiveSupport::Notifications::Event.new(*args)
+  class_name = event.payload[:class_name]
+  record_count = event.payload[:record_count]
+  Foreman::Telemetry.increment "instantiation.#{class_name}"
+end
```

Comment:
This probe might be too much, I need to test on an instance with some load. Maybe we can do 10% sampling to reduce load, it will be very interesting to see if we don't make too many instances of DB records somewhere.